src/test/java/DirectionsServletTest.java

```diff
@@ -0,0 +1,189 @@
+          "[DirectionsLeg: \"D\" -> \"A\" (42,-83 -> 45,-73), duration=9 hours, distance=8 km: 2 steps]");
+  private static final List<String> A_TO_B_NO_WAYPOINTS =
+      ImmutableList.of(
+          "[DirectionsLeg: \"A\" -> \"B\" (45,-73 -> 43,-80), duration=6 hours, distance=6 km: 2 steps]");
+
+  @Before
+  public void setUp() throws IOException {
+    request = Mockito.mock(HttpServletRequest.class);
+    response = Mockito.mock(HttpServletResponse.class);
+    directionsClientFactory = Mockito.mock(DirectionsClientFactory.class);
+    directionsClient = Mockito.mock(DirectionsClient.class);
+    // Writer used in get/post requests to capture HTTP response values
+    stringWriter = new StringWriter();
+    printWriter = new PrintWriter(stringWriter);
+    gson = new Gson();
+    Mockito.when(directionsClientFactory.getDirectionsClient(API_KEY)).thenReturn(directionsClient);
+    Mockito.when(response.getWriter()).thenReturn(printWriter);
+  }
+
+  @Test
+  public void aToBWithWaypoints() throws DirectionsException, ServletException {
+    servlet =
+        new DirectionsServlet(
+            directionsClientFactory, API_KEY, "A", "B", ImmutableList.of("C", "D"));
+
+    Mockito.when(directionsClient.getDirections("A", "B", ImmutableList.of("C", "D")))
+        .thenReturn(A_TO_B_WITH_WAYPOINTS);
+
+    servlet.doGet(request, response);
+    printWriter.flush();
+
+    String actualString = stringWriter.toString();
+    Type type = new TypeToken<List<String>>() {}.getType();
+    List<String> actual = gson.fromJson(actualString, type);
+
+    Assert.assertEquals(A_TO_B_WITH_WAYPOINTS, actual);
+  }
+
+  @Test
+  public void aToAWithWaypoints() throws DirectionsException, ServletException {
+    servlet =
+        new DirectionsServlet(
+            directionsClientFactory, API_KEY, "A", "A", ImmutableList.of("B", "C", "D"));
+
+    Mockito.when(directionsClient.getDirections("A", "A", ImmutableList.of("B", "C", "D")))
+        .thenReturn(A_TO_A_WITH_WAYPOINTS);
+
+    servlet.doGet(request, response);
+    printWriter.flush();
+
+    String actualString = stringWriter.toString();
+    Type type = new TypeToken<List<String>>() {}.getType();
+    List<String> actual = gson.fromJson(actualString, type);
+
+    Assert.assertEquals(A_TO_A_WITH_WAYPOINTS, actual);
+  }
+
+  @Test
+  public void aToBNoWaypoints() throws DirectionsException, ServletException {
+    servlet = new DirectionsServlet(directionsClientFactory, API_KEY, "A", "B", ImmutableList.of());
+
+    Mockito.when(directionsClient.getDirections("A", "B", ImmutableList.of()))
+        .thenReturn(A_TO_B_NO_WAYPOINTS);
+
+    servlet.doGet(request, response);
+    printWriter.flush();
+
+    String actualString = stringWriter.toString();
+    Type type = new TypeToken<List<String>>() {}.getType();
+    List<String> actual = gson.fromJson(actualString, type);
+
+    Assert.assertEquals(A_TO_B_NO_WAYPOINTS, actual);
+  }
+
+  @Test
+  public void aToBNoWaypointsParseResult()
```

Comment:
I think this test should be in a separate `DirectionClientImplTest` class?

Reply:
Agreed, thanks!

Reply:
`RESOLVED`